Add variants for extensions without main hosts

dennisvandehoef opened the pull request:
This pull request adds extra variants for extensions without the unified hosts file.

I do realize that in https://github.com/StevenBlack/hosts/issues/2328 I suggested to use `alternates/porn/reduced` as the filename, though this was not able without massively refactoring all the code. As most code is written in a way to only have one hosts-file per directory. Therefore, I used `alternates/reduced-porn/hosts`.

This PR closes https://github.com/StevenBlack/hosts/issues/2328, https://github.com/StevenBlack/hosts/issues/1429 and maybe https://github.com/StevenBlack/hosts/issues/1954. It also offers a better solution for the already closed https://github.com/StevenBlack/hosts/issues/358.

StevenBlack commented:
Thank you for this extensive Pull Request Dennis @dennisvandehoef.

I'll me leaving many review comments.

For now I have a question: given we separate the component categories now, is it not sensible to combine those in various ways, too? 

I'm open to having it either way, but since everything else here is an exercise in combinatorics...

StevenBlack commented:
Dennis @dennisvandehoef it's important to be clear: the only three files that matter here are 
* readme_template.md
* updateHostsFile.py
* updateReadMe.py

The other 26 files are all generated from those. In some sense, those pollute this pull request. Or it feels that way, a little.

dennisvandehoef commented on updateHostsFile.py:
```diff
@@ -1424,21 +1465,25 @@
 
     # Create new empty hosts file
-    open(old_file_path, "a").close()
+    open(full_file_path, "a").close()
```
I wonder how an open and direct close, can result in sometimes staying open 😕

dennisvandehoef replied:
I think this is a bug in CodeQL. If I look at the [documentation of this warning](https://codeql.github.com/codeql-query-help/python/py-file-not-closed/) it seems to not expect creating a file without using it as we do here.

We have no operations between open and close, and therefore there is no error that can prevent us from closing after we opened.

dennisvandehoef commented:
Heey @StevenBlack 

Thanks for the quick reply.

> given we separate the component categories now, is it not sensible to combine those in various ways, too?

I was also doubting if it would be good to add the other combinations. Most programs I checked can load multiple hosts files, so with these versions the users can already combine every combination they want. Nevertheless, I also realize that https://github.com/StevenBlack/hosts/issues/1954 explicitly asked for this combination. I will add the other combinations as well.

> the only three files that matter here are readme_template.md, updateHostsFile.py, updateReadMe.py. The other 26 files are all generated from those.

You forgot makeHosts.py 😉  On a more serious note, I am aware and did not know if I should run the changes and include the output or not. It felt strange to not add the new hosts-files to the MR, and by that I also had to push all other README files that now are expected to link them. I see I need to rebase anyway, and will split this into two commits (code changes and generator output) to make review easier. What do you think?

StevenBlack commented:
Dennis @dennisvandehoef you're right about makehostfiles.py — the one I use all the time 🙃

Upon reflection, leave the generated files in the PR. After all we have to view the readme files to ensure no glitches...

StevenBlack commented:
Dennis @dennisvandehoef of course let's ignore the `readme.md` and `readmeData.json` merge conflicts for the duration of this PR.

dennisvandehoef commented:
Hey @StevenBlack,

The CI pipeline should pass now, at least do all tests run through locally. I also added some tests for this feature.

Also, as a bonus, I changed `makeHosts.py` in a way that we can easily add more extensions and can make sure that we always generate the version with and without the unified hosts for all combinations of extensions. In the future, if a new extension will be added, we only need to add it to the end of the array, and all new combinations will be automatically taken care of by the script.

And last but not least, I moved the generated data to a different commit, to make reviewing easier 😸

StevenBlack commented:
Dennis @dennisvandehoef instead of `reduced` in folder names like `[alternates/reduced-gambling-porn-social/` can we use the suffix `-only` instead?  I think that reads better, and ultimately requires less explanation.

In other words, change...
```
alternates/reduced-gambling-porn-social/
```
...to...
```
alternates/gambling-porn-social-only/
```

dennisvandehoef commented on updateHostsFile.py:
```diff
@@ -677,13 +690,14 @@
+        for source in sort_sources(
+            recursive_glob(sources_params["datapath"], source_data_filename)
+        ):
+            update_file = open(source, "r", encoding="UTF-8")
```
This time the tool was right, though it is now fixed.

dennisvandehoef commented:
Hey @StevenBlack 

I really like the suggestion, that name is way better.
Changed it 👍

StevenBlack commented:
Dennis @dennisvandehoef sorry for the delay in fully evaluating this, I'm overcome by events 😄 but I will test all this soon.  Thank you for your patience.

StevenBlack commented:
## Thank you Dennis van de Hoef @dennisvandehoef!!

You did a great job, and yeoman service, with this pull request  👊🏻💥❗  Thank you for your patience because it took me a long time to merge this.

I have cherry-picked the changes to the relevant python files and templates, and added those to `master` branch. After generating the various hosts products, including the new ones. and updating stats, I am proud to [release Hosts version 3.13.0](https://github.com/StevenBlack/hosts/releases/tag/3.13.0)!

Here's a link to our [updated list of contributors](https://github.com/StevenBlack/hosts/graphs/contributors).

This is awesome!  THANK YOU DENNIS @dennisvandehoef!

Closing.
